### PR TITLE
Improvements

### PR DESCRIPTION
**Pull request by amerkoleci**
ApplicationInfo ApplicationVersion and EngineVersion are encoded Version int.
Add implicit conversion from Version to int.
Make VulkanLibrary public for external call extendibility.
Some UnmanagedFunctionPointer WinApi.
Add new ClearColorValue constructor taking r, g, b, a.

### CODE REVIEW

**Comment by amerkoleci** [comment]
If you want you can merge thoose improvements, I will add more later on

**Comment by WardBenjamin** [comment]
I appreciate the Version changes - been using reflection to get at that value but I'd forgotten to open a PR.

File: Src/AllocationCallbacks.cs [hunk 1]
```diff
@@ -144,6 +144,7 @@ internal void ToNative(Native* native)
         /// <param name="allocationScope">
         /// Value specifying the allocation scope of the lifetime of the allocation.
         /// </param>
+        [UnmanagedFunctionPointer(CallingConvention.Winapi)]
```
Comment (discosultan) [inline Src/AllocationCallbacks.cs]: Why is this needed?

File: Src/CommandBuffer.cs [hunk 2]
```diff
@@ -1963,6 +1963,18 @@ public ClearColorValue(ColorU4 value) : this()
...
         }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ClearColorValue"/> structure.
+        /// </summary>
+        /// <param name="r">The red clear value.</param>
+        /// <param name="g">The green clear value.</param>
+        /// <param name="b">The blue clear value.</param>
+        /// <param name="a">The alpha clear value.</param>
+        public ClearColorValue(float r, float g, float b, float a = 1.0f) : this()
```
Comment (discosultan) [inline Src/CommandBuffer.cs]: Great addition 👍

File: Src/Ext/DebugReportCallbackExt.cs [hunk 3]
```diff
@@ -35,7 +35,7 @@ public unsafe class DebugReportCallbackExt : DisposableHandle<long>
                     });
                 callbackHandle = Interop.GetFunctionPointerForDelegate(_callback);
             }
-                        createInfo.ToNative(out DebugReportCallbackCreateInfoExt.Native nativeCreateInfo, callbackHandle);
+            createInfo.ToNative(out DebugReportCallbackCreateInfoExt.Native nativeCreateInfo, callbackHandle);
```
Comment (discosultan) [inline Src/Ext/DebugReportCallbackExt.cs]: 👍

File: Src/Instance.cs [hunk 4]
```diff
@@ -307,7 +313,7 @@ public unsafe struct ApplicationInfo
         /// The unsigned integer variable containing the developer-supplied version
         /// number of the application.
         /// </summary>
-        public int ApplicationVersion;
+        public Version ApplicationVersion;
```
Comment (discosultan) [inline Src/Instance.cs]: I never realized these values were using the `Version` encoding scheme.

File: Src/Version.cs [hunk 5]
```diff
@@ -62,5 +62,10 @@ public Version(int major, int minor, int patch)
         /// A shorthand for writing <c>new Version(0, 0, 0)</c>.
         /// </summary>
         public static Version Zero => new Version(0, 0, 0);
+
+        public static implicit operator int(Version version)
```
Comment (discosultan) [inline Src/Version.cs]: Where is this implicit conversion useful?

File: Src/VulkanLibrary.cs [hunk 6]
```diff
@@ -5,7 +5,10 @@
 
 namespace VulkanCore
 {
-    internal static class VulkanLibrary
+    /// <summary>
+    /// Static vulkan library loaded from host platform.
+    /// </summary>
+    public static class VulkanLibrary
```
Comment (discosultan) [inline Src/VulkanLibrary.cs]: That's a good idea to make it public if it's useful outside this library.

File: Src/Version.cs (hunk 5 above)
Comment (amerkoleci) [reply Src/Version.cs]: Doesn't hurt and for debug purpose is good

File: Src/AllocationCallbacks.cs (hunk 1 above)
Comment (amerkoleci) [reply Src/AllocationCallbacks.cs]: The vulkan call is stdcall on windows and cdecl on non windows

Comment (discosultan) [reply Src/AllocationCallbacks.cs]: What's the default behavior?

Comment (amerkoleci) [reply Src/AllocationCallbacks.cs]: None, if you dont mark without UnmanagedFunctionPointer, WinApi behave as I described,  StdCall on windows Cdecl else where

**Comment by amerkoleci** [comment]
Can we get this merged? I would love to work on 1.1 additions.

**Comment by discosultan** [comment]
@amerkoleci Yes, let's get this resolved today. I just have a few additional questions regarding the `UnmanagedFunctionPointer` usage.

File: Src/CommandPool.cs [hunk 7]
```diff
@@ -85,12 +85,15 @@ public override void Dispose()
             base.Dispose();
         }
 
+        [UnmanagedFunctionPointer(CallingConvention.Winapi)]
```
Comment (discosultan) [inline Src/CommandPool.cs]: 1. Can you describe a use case where not having this attribute is an issue? I'm just trying to get a better overview of the problem here.

2. Why is the attribute applied only to those couple of delegate types in `AllocationCallbacks`, `CommandPool` and `Device` classes?

Comment (amerkoleci) [reply Src/CommandPool.cs]: 1) I've added that as Mono Vulkan C# (https://github.com/mono/VulkanSharp/blob/master/src/Vulkan/Interop/ImportedCommands.cs), dunno what is the default calling convension.
2) Didn't had time to add on rest of delegates.

Comment (discosultan) [reply Src/CommandPool.cs]: Alright. Before I could merge these kind of changes, I would need to get a clear understanding of what problem it solves. I wouldn't just blindly follow other implementations without reasoning. Also, if it's going into master branch, it shouldn't be just a partial fix, but should be applied everywhere.

Could you remove those attributes from this PR and I'd happily merge rest of the changes. We can tackle the `UnmanagedFunctionPointer` in a separate issue or PR if you're interested.

Thanks!

Comment (amerkoleci) [reply Src/CommandPool.cs]: Okay, will do ASAP

**Comment by discosultan** [comment]
Thanks 👍

**Comment by WardBenjamin** [comment]
Solid, thank you!
